x-pack/heartbeat/monitors/browser/source/zipurl.go

```diff
@@ -0,0 +1,236 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package source
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type ZipURLSource struct {
+	URL      string `config:"url" json:"url"`
+	Folder   string `config:"folder" json:"folder"`
+	Username string `config:"username" json:"username"`
+	Password string `config:"password" json:"password"`
+	Retries  int    `config:"retries" default:"3" json:"retries"`
+	BaseSource
+	// Etag from last successful fetch
+	etag            string
+	TargetDirectory string `config:"target_directory" json:"target_directory"`
+}
+
+var ErrNoEtag = fmt.Errorf("No ETag header in zip file response. Heartbeat requires an etag to efficiently cache downloaded code")
+
+func (z *ZipURLSource) Fetch() error {
+	changed, err := checkIfChanged(z)
+	if err != nil {
+		return fmt.Errorf("could not check if zip source changed for %s: %w", z.URL, err)
+	}
+	if !changed {
+		return nil
+	}
+
+	// remove target directory if etag changed
+	if z.TargetDirectory != "" {
+		os.RemoveAll(z.TargetDirectory)
+	}
+
+	tf, err := ioutil.TempFile("/tmp", "elastic-synthetics-zip-")
+	if err != nil {
+		return fmt.Errorf("could not create tmpfile for zip source: %w", err)
+	}
+	defer os.Remove(tf.Name())
+
+	newEtag, err := download(z, tf)
+	if err != nil {
+		return fmt.Errorf("could not download %s: %w", z.URL, err)
+	}
+	// We are guaranteed an etag
+	z.etag = newEtag
+
+	if z.TargetDirectory != "" {
+		os.MkdirAll(z.TargetDirectory, 0755)
+	} else {
+		z.TargetDirectory, err = ioutil.TempDir("/tmp", "elastic-synthetics-unzip-")
+		if err != nil {
+			return fmt.Errorf("could not make temp dir for zip download: %w", err)
+		}
+	}
+
+	err = unzip(tf, z.TargetDirectory, z.Folder)
+	if err != nil {
+		z.Close()
+		return err
+	}
+
+	// run as the local job after extracting the files
+	if !Offline() {
+		err = setupOnlineDir(z.TargetDirectory)
+		if err != nil {
+			os.RemoveAll(z.TargetDirectory)
+			return fmt.Errorf("failed to install dependencies at: '%s' %w", z.TargetDirectory, err)
+		}
+	}
+
+	return nil
+}
+
+func unzip(tf *os.File, targetDir string, folder string) error {
+	stat, err := tf.Stat()
+	if err != nil {
+		return err
+	}
+
+	rdr, err := zip.NewReader(tf, stat.Size())
+	if err != nil {
+		return fmt.Errorf("could not read file %s as zip: %w", tf.Name(), err)
+	}
+
+	for _, f := range rdr.File {
+		err = unzipFile(targetDir, folder, f)
+		if err != nil {
+			// TODO: err handlers
+			os.RemoveAll(targetDir)
```

Comment:
This feels unnecessary right? we already remove the directory when unzip returns error. 

https://github.com/elastic/beats/pull/24714/files#diff-a42b602f839672a041979a73130887f65c77d4792a207ce86444af8a01660012R71